### PR TITLE
feat(command): ListTaskCommand - display substasks

### PR DESCRIPTION
**Pull request by jvancoillie**
| Q                | A
| ---------------- | ---
| PHP version?     | 7.4
| Bundle version?  | -
| New feature?     | yes (improvement)
| Bug fix?         | no
| Discussion?      | #87

Display related substasks and type task in `scheduler:list`command.

output: 
```bash 


 [OK] 3 tasks found


+-------------+---------+------------------------+------------+---------------------------+---------------------------+-------------------------+-----------------------------+---------+------+
| Type        | Name    | Description            | Expression | Last execution date       | Next execution date       | Last execution duration | Last execution memory usage | State   | Tags |
+-------------+---------+------------------------+------------+---------------------------+---------------------------+-------------------------+-----------------------------+---------+------+
| ChainedTask | chained | chained with 2 tasks   | * * * * *  | 2021-05-22T16:24:48+00:00 | 2021-05-24T10:31:00+00:00 | < 1 sec                 | 23.3 MiB                    | enabled |      |
|           > | foo     | command foo chained    | -          | Not executed              | 2021-05-24T10:31:00+00:00 | Not tracked             | Not tracked                 | enabled |      |
|           > | bar     | command bar in chained | -          | 2021-05-22T16:24:48+00:00 | 2021-05-24T10:31:00+00:00 | < 1 sec                 | 23.3 MiB                    | enabled |      |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CommandTask | task5   | command task 5         | * * * * *  | 2021-05-22T16:24:48+00:00 | 2021-05-24T10:31:00+00:00 | < 1 sec                 | 23.4 MiB                    | enabled |      |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CommandTask | task6   | command task 6         | * * * * *  | 2021-05-22T16:24:48+00:00 | 2021-05-24T10:31:00+00:00 | < 1 sec                 | 23.4 MiB                    | enabled |      |
+-------------+---------+------------------------+------------+---------------------------+---------------------------+-------------------------+-----------------------------+---------+------+
```

### CODE REVIEW

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Could you rebase this PR against `main` and see if the code can be improved thanks to recent changes? 

Thanks 🙂

File: tests/Command/ListTasksCommandTest.php [hunk 1]
```diff
@@ -116,6 +119,61 @@ public function testCommandCanListTaskWithSpecificState(string $stateOption): vo
...
+        $secondTask->expects(self::once())->method('getDescription')->willReturn('A bar task');
+        $secondTask->expects(self::once())->method('getExpression')->willReturn('* * * * *');
+        $secondTask->expects(self::exactly(2))->method('getLastExecution')->willReturn(new DateTimeImmutable());
+        $secondTask->expects(self::once())->method('getState')->willReturn(TaskInterface::ENABLED);
+        $secondTask->expects(self::once())->method('getTags')->willReturn(['app', 'slow']);
+
+        $chainedTask = new ChainedTask('nested');
+        $chainedTask->setDescription('A nested task');
+        $chainedTask->setLastExecution(new DateTimeImmutable());
+        $chainedTask->setTasks(...[$secondTask, $task]);
```
Comment (Guikingone) [inline tests/Command/ListTasksCommandTest.php:143]: Could be `$chainedTask->setTasks($secondTask, $task);`

File: src/Command/ListTasksCommand.php [hunk 2]
```diff
@@ -102,20 +105,44 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
-            null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed',
-            (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM),
-            null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked',
-            null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked',
-            $task->getState(),
-            implode(', ', $task->getTags()) ?? 'No tags set',
-        ], $tasks));
+        $table->setHeaders(['Type', 'Name', 'Description', 'Expression',  'Last execution date', 'Next execution date', 'Last execution duration', 'Last execution memory usage', 'State', 'Tags']);
+        $first = true;
+        foreach ($tasks as $task) {
```
Comment (Guikingone) [inline src/Command/ListTasksCommand.php:111]: Could we use `array_walk`?

Comment (jvancoillie) [reply src/Command/ListTasksCommand.php]: thanks for this review, push new version.

File: src/Command/ListTasksCommand.php [hunk 3]
```diff
@@ -102,20 +105,39 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
-            null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed',
-            (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM),
-            null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked',
-            null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked',
-            $task->getState(),
-            implode(', ', $task->getTags()) ?? 'No tags set',
-        ], $tasks));
+        $table->setHeaders(['Type', 'Name', 'Description', 'Expression',  'Last execution date', 'Next execution date', 'Last execution duration', 'Last execution memory usage', 'State', 'Tags']);
+
+        array_walk($tasks, function (TaskInterface $task) use ($table): void {
```
Comment (Guikingone) [inline src/Command/ListTasksCommand.php:111]: If I'm right, we can use:

```php
$tasks->walk(function (TaskInterface $task) use ($table): void {
    // 
});
```

This is probably a better idea as we can iterate over the list pointer and not create a new one here 🙂

File: src/Command/ListTasksCommand.php [hunk 4]
```diff
@@ -102,20 +105,39 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+                null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed',
+                (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM),
+                null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked',
+                null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked',
+                $task->getState(),
+                implode(', ', $task->getTags()) ?? 'No tags set',
+            ]);
+
+            if ($task instanceof ChainedTask) {
+                $table->addRows(array_map(fn (TaskInterface $task): array => [
```
Comment (Guikingone) [inline src/Command/ListTasksCommand.php:126]: I wonder if it could be a good idea to add a method `map()` inside `TaskList`, this way, we can iterate directly via the list without importing the method and/or leaving the list, any thought? 🤔

**Comment by Guikingone** [comment]
BTW, nice work on this one @jvancoillie 🚀

File: src/Command/ListTasksCommand.php (hunk 3 above)
Comment (jvancoillie) [reply src/Command/ListTasksCommand.php:111]: you're right, I completely forgot about this new function: D

File: src/Command/ListTasksCommand.php (hunk 4 above)
Comment (jvancoillie) [reply src/Command/ListTasksCommand.php:126]: yes I think it's a good idea, in any case we already have a walk method so why not a map method

Comment (jvancoillie) [reply src/Command/ListTasksCommand.php:123]: Hello @Guikingone 👋 

After some thought, I wonder if implementing the `map ()` method is really a good idea because it won't be able to return a `TaskListInterface` but a simple array.
What will be the added value of this method ?

WDYT ?

Comment (Guikingone) [reply src/Command/ListTasksCommand.php:123]: Hi 👋🏻 

IMHO, that's not an issue to worry about, the `TaskListInterface` can either return a list, an array or even an integer, that not something that lead to an issue.

The value is mainly the developer experience as we often use `array_map` on the list in the bundle and personally, I prefer to use `list->map(...)` rather than `array_map($list->toArray()...)`, in the end, the behaviour still pretty must the same but the first syntax is "better" to read 🤔

Comment (jvancoillie) [reply src/Command/ListTasksCommand.php:123]: it's ok for me, i will work on it

**Comment by jvancoillie** [comment]
Here are the changes I made what do you think?

**Comment by Guikingone** [comment]
Yes, I'll take a look at this tonight, thanks for the work 🙂  👍🏻

File: src/Task/TaskList.php [hunk 5]
```diff
@@ -163,6 +163,14 @@ public function walk(Closure $func): TaskListInterface
         return $this;
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function map(Closure $func): array
+    {
+        return array_map($func, $this->tasks);
```
Comment (Guikingone) [inline src/Task/TaskList.php:171]: Could you add the import instruction at the top of the file?

File: tests/Command/ListFailedTasksCommandTest.php [hunk 6]
```diff
@@ -32,7 +33,7 @@ public function testCommandIsConfigured(): void
     public function testCommandCannotListEmptyFailedTasks(): void
     {
         $failedTasks = $this->createMock(TaskListInterface::class);
-        $failedTasks->expects(self::once())->method('toArray')->willReturn([]);
+        $failedTasks->expects(self::once())->method('count')->willReturn(0);
```
Comment (Guikingone) [inline tests/Command/ListFailedTasksCommandTest.php:36]: I think we can instantiate directly a `TaskList` here, no need to mock it 🙂

File: tests/Command/ListTasksCommandTest.php [hunk 7]
```diff
@@ -55,7 +56,7 @@ public function testCommandIsCorrectlyConfigured(): void
     public function testCommandCannotReturnTaskOnEmptyScheduler(): void
     {
         $taskList = $this->createMock(TaskListInterface::class);
-        $taskList->expects(self::once())->method('toArray')->willReturn([]);
+        $taskList->expects(self::once())->method('count')->willReturn(0);
```
Comment (Guikingone) [inline tests/Command/ListTasksCommandTest.php:59]: Same

File: tests/Task/LazyTaskListTest.php [hunk 8]
```diff
@@ -119,6 +119,23 @@ public function testListCanWalkThroughTask(): void
...
+        self::assertSame(['foo' => 'foo', 'bar' => 'bar'], $list->map(fn (TaskInterface $task) => $task->getName()));
+    }
+
+    public function testNotInitializedListCanApplyMapClosure(): void
+    {
+        $list = new LazyTaskList(new TaskList([new NullTask('foo'), new NullTask('bar')]));
+
+        self::assertSame(['foo' => 'foo', 'bar' => 'bar'], $list->map(fn (TaskInterface $task) => $task->getName()));
+    }
+
```
Comment (Guikingone) [inline tests/Task/LazyTaskListTest.php:138]: Could we add some assertions on the `isInitialized()` method? 

The lazy list should not be initialised before called calling the method then initialised right after it.

File: tests/Task/TaskListTest.php [hunk 9]
```diff
@@ -252,4 +252,14 @@ public function testListCanApplyClosureOnEachTask(): void
         self::assertCount(1, $nullTask->getTags());
         self::assertContains('walk', $nullTask->getTags());
     }
+
+    public function testListCanApplyMapClosure(): void
+    {
+        $firstNullTask = new NullTask('foo');
+        $secondNullTask = new NullTask('bar');
+
+        $taskList = new TaskList([$firstNullTask, $secondNullTask]);
```
Comment (Guikingone) [inline tests/Task/TaskListTest.php:261]: Maybe just: 

```php
$taskList = new TaskList([
    new NullTask('foo'), 
    new NullTask('bar'),
]);
```

File: tests/Task/TaskListTest.php [hunk 10]
```diff
@@ -252,4 +252,14 @@ public function testListCanApplyClosureOnEachTask(): void
...
+
+        self::assertSame(['foo' => 'foo', 'bar' => 'bar'], $taskList->map(fn (TaskInterface $task) => $task->getName()));
```
Comment (Guikingone) [inline tests/Task/TaskListTest.php:263]: Could we specify the return type of this closure?

File: tests/Task/LazyTaskListTest.php [hunk 11]
```diff
@@ -119,6 +119,23 @@ public function testListCanWalkThroughTask(): void
...
+        $list->add(new NullTask('bar'));
+
+        self::assertSame(['foo' => 'foo', 'bar' => 'bar'], $list->map(fn (TaskInterface $task) => $task->getName()));
+    }
+
+    public function testNotInitializedListCanApplyMapClosure(): void
+    {
+        $list = new LazyTaskList(new TaskList([new NullTask('foo'), new NullTask('bar')]));
+
+        self::assertSame(['foo' => 'foo', 'bar' => 'bar'], $list->map(fn (TaskInterface $task) => $task->getName()));
```
Comment (Guikingone) [inline tests/Task/LazyTaskListTest.php:136]: Could we specify the return type of this closure?

File: tests/Task/LazyTaskListTest.php [hunk 12]
```diff
@@ -119,6 +119,23 @@ public function testListCanWalkThroughTask(): void
...
     }
 
+    public function testInitializedListCanApplyMapClosure(): void
+    {
+        $list = new LazyTaskList(new TaskList());
+
+        $list->add(new NullTask('foo'));
+        $list->add(new NullTask('bar'));
+
+        self::assertSame(['foo' => 'foo', 'bar' => 'bar'], $list->map(fn (TaskInterface $task) => $task->getName()));
```
Comment (Guikingone) [inline tests/Task/LazyTaskListTest.php:129]: Could we specify the return type of this closure?

**Comment by jvancoillie** [comment]
Thanks for this review you are awesome.  I will work on this

File: tests/Task/LazyTaskListTest.php (hunk 8 above)
Comment (jvancoillie) [reply tests/Task/LazyTaskListTest.php:140]: This method is not available, should I add it?

Comment (jvancoillie) [reply tests/Task/LazyTaskListTest.php:140]: sorry, i will rebase main :(

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Looks good to me, could you improve tests as reported by Infection? 

This way, we can merge this improvement for `0.5.0` 🙂

**Comment by jvancoillie** [comment]
Hello @Guikingone 

> Looks good to me, could you improve tests as reported by Infection?

I am not familiar with the infection report, can you give me some ideas for improvement or information to see more clearly

**Comment by Guikingone** [comment]
> I am not familiar with the infection report, can you give me some ideas for improvement or information to see more clearly

Sure, here's the points you can probably work on: 

**First:**

```bash
39) /home/runner/work/SchedulerBundle/SchedulerBundle/src/Command/ListTasksCommand.php:83    [M] DecrementInteger

--- Original
+++ New
@@ @@
     {
         $symfonyStyle = new SymfonyStyle($input, $output);
         $tasks = $this->scheduler->getTasks();
-        if (0 === $tasks->count()) {
+        if (-1 === $tasks->count()) {
             $symfonyStyle->warning('No tasks found');
             return self::SUCCESS;
         }
```

The issue here is the integer usage, maybe you can add a test where the `$tasks` is not empty and test if the warning is displayed. 

**Second:**

```bash
42) /home/runner/work/SchedulerBundle/SchedulerBundle/src/Command/ListTasksCommand.php:122    [M] ArrayItemRemoval

--- Original
+++ New
@@ @@
         $tasks->walk(function (TaskInterface $task) use($table) : void {
             $table->addRow([(new ReflectionClass($task))->getShortName(), $task->getName(), $task->getDescription() ?? 'No description set', $task->getExpression(), null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed', (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM), null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked', null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked', $task->getState(), implode(', ', $task->getTags()) ?? 'No tags set']);
             if ($task instanceof ChainedTask) {
-                $table->addRows($task->getTasks()->map(fn(TaskInterface $task): array => ['          >', $task->getName(), $task->getDescription() ?? 'No description set', '-', null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed', (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM), null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked', null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked', $task->getState(), implode(', ', $task->getTags()) ?? 'No tags set']));
+                $table->addRows($task->getTasks()->map(fn(TaskInterface $task): array => [$task->getName(), $task->getDescription() ?? 'No description set', '-', null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed', (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM), null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked', null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked', $task->getState(), implode(', ', $task->getTags()) ?? 'No tags set']));
             }
         });
         $table->render();
```

The issue here is related to the `>` that you use to display a sub-task, I'm not sure that you've added a test that validate this behaviour. 

**Third:**

```bash
43) /home/runner/work/SchedulerBundle/SchedulerBundle/src/Command/ListTasksCommand.php:129    [M] NotIdentical

--- Original
+++ New
@@ @@
         $tasks->walk(function (TaskInterface $task) use($table) : void {
             $table->addRow([(new ReflectionClass($task))->getShortName(), $task->getName(), $task->getDescription() ?? 'No description set', $task->getExpression(), null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed', (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM), null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked', null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked', $task->getState(), implode(', ', $task->getTags()) ?? 'No tags set']);
             if ($task instanceof ChainedTask) {
-                $table->addRows($task->getTasks()->map(fn(TaskInterface $task): array => ['          >', $task->getName(), $task->getDescription() ?? 'No description set', '-', null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed', (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM), null !== $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked', null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked', $task->getState(), implode(', ', $task->getTags()) ?? 'No tags set']));
+                $table->addRows($task->getTasks()->map(fn(TaskInterface $task): array => ['          >', $task->getName(), $task->getDescription() ?? 'No description set', '-', null !== $task->getLastExecution() ? $task->getLastExecution()->format(DATE_ATOM) : 'Not executed', (new CronExpression($task->getExpression()))->getNextRunDate()->format(DATE_ATOM), null === $task->getExecutionComputationTime() ? Helper::formatTime($task->getExecutionComputationTime() / 1000) : 'Not tracked', null !== $task->getExecutionMemoryUsage() ? Helper::formatMemory($task->getExecutionMemoryUsage()) : 'Not tracked', $task->getState(), implode(', ', $task->getTags()) ?? 'No tags set']));
             }
         });
         $table->render();
```

Pretty similar but regarding the `getExecutionComputationTime` method, it may be a good idea to add a test that valide the display of the value. 

It should be enough to make the CI goes green for now 🙂

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

Nice PR, a huge thank on this one, the feature will be released via `0.5.0` 🙂
